lib/improver/nbhood.py

```diff
@@ -431,21 +433,44 @@ def run(cube, radius):
             Cube containing the smoothed field after the square neighbourhood
             method has been applied.
         """
-        if isinstance(cube.data, np.ma.MaskedArray):
-            msg = ('Masked data is not currently supported in ',
-                   'SquareNeighbourhood.')
-            raise ValueError(msg)
+        masked_flag = False
         if np.any(np.isnan(cube.data)):
             msg = ('Data array contains NaNs which are not currently ',
                    'supported in SquareNeighbourhood.')
             raise ValueError(msg)
-        summed_up_cube = SquareNeighbourhood.cumulate_array(cube)
-        grid_cells_x, grid_cells_y = (
-            Utilities.get_neighbourhood_width_in_grid_cells(
-                summed_up_cube, radius, MAX_RADIUS_IN_GRID_CELLS))
-        neighbourhood_averaged_cube = (
-            SquareNeighbourhood.mean_over_neighbourhood(
-                summed_up_cube, grid_cells_x, grid_cells_y))
+        # If the data is masked, the mask will be processed as well as the
+        # original_data * mask array.
+        if isinstance(cube.data, np.ma.MaskedArray):
+            masked_flag = True
+            mask_cube = cube.copy()
+            mask_cube.data = np.logical_not(cube.data.mask.astype(int))
+            cube.data = cube.data.data * mask_cube.data
+            cubes_to_sum = iris.cube.CubeList([cube, mask_cube])
+        else:
+            cubes_to_sum = [cube]
+        neighbourhood_averaged_cubes = []
+        for cube in cubes_to_sum:
```

Comment:
Please don't use the name cube in this loop, as it overwrites the reference to the original input cube.